### PR TITLE
Homework solution

### PR DESCRIPTION
**Pull request by ohatsenko**

### CODE REVIEW

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 1]
```diff
@@ -2,13 +2,67 @@
 
 import core.basesyntax.dao.StorageDao;
 import core.basesyntax.dao.StorageDaoImpl;
+import core.basesyntax.exception.RegistrationException;
 import core.basesyntax.model.User;
 
 public class RegistrationServiceImpl implements RegistrationService {
+    private static final int MIN_AGE = 18;
+    private static final int MIN_LENGTH = 6;
+
```
Comment (irynamekh) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:11]: extra empty line

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 2]
```diff
@@ -2,13 +2,67 @@
...
     private final StorageDao storageDao = new StorageDaoImpl();
 
     @Override
     public User register(User user) {
-        return null;
+        checkUser(user);
+        checkLogin(user);
+        checkPassword(user);
+        checkAge(user);
+
```
Comment (irynamekh) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:20]: same

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+import org.junit.jupiter.api.Test;
+
+class RegistrationServiceImplTest {
+    private static final User NULL_USER = null;
+    private static final String NULL_LOGIN = null;
+    private static final Integer NULL_AGE = null;
+    private static final String NULL_PASSWORD = null;
+    private static final String VALID_LOGIN1 = "Andrew";
+    private static final String VALID_LOGIN2 = "Penelopa";
+
```
Comment (irynamekh) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:19]: same

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    @Test
+    void register_twoUsers_Ok() {
+        User validUser2 = regService.register(VALID_USER2);
+        User validUser3 = regService.register(VALID_USER3);
+        assertEquals(validUser2, VALID_USER2);
+        assertEquals(validUser3, VALID_USER3);
+    }
+
```
Comment (irynamekh) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:133]: it seems to me that there should be another method that is executed at the end of all tests and cleans the storage.

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 5]
```diff
@@ -2,13 +2,67 @@
...
+        if (user.getAge() == null) {
+            throw new RegistrationException("Age can't be null");
+        }
+        if (user.getAge() < MIN_AGE) {
+            throw new RegistrationException("Not valid age: "
+                    + user.getAge()
+                    + ". Min allowed age is " + MIN_AGE);
+        }
+    }
+
```
Comment (irynamekh) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:66]: same

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 6]
```diff
@@ -2,13 +2,67 @@
...
+    private void checkPassword(User user) {
+        if (user.getPassword() == null) {
+            throw new RegistrationException("Password can't be null");
+        }
+        if (user.getPassword().length() < MIN_LENGTH) {
+            throw new RegistrationException("Not valid password length: "
+                    + user.getPassword().length()
+                    + ". Min allowed password length is " + MIN_LENGTH);
+        }
+
```
Comment (irynamekh) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:53]: same

**Review by irynamekh — CHANGES_REQUESTED** [review]

**Review by irynamekh — APPROVED** [review]

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    private static final String VALID_LOGIN1 = "Andrew";
+    private static final String VALID_LOGIN2 = "Penelopa";
+    private static final String VALID_PWD = "1234567";
+    private static final int VALID_AGE = 20;
+    private static final String INVALID_LOGIN = "Andr";
+    private static final String INVALID_PWD = "12345";
+    private static final int INVALID_AGE = 17;
+    private static final int MIN_AGE = 18;
+    private static final int MIN_LENGTH = 6;
+    private static final User INVALID_LOGIN_USER = new User(INVALID_LOGIN, VALID_PWD, VALID_AGE);
```
Comment (andrii-hoienko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:27]: Please do not hardcode users, it is better to create them in test methods using constuctor/factory method/builder

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    private static final User INVALID_PWD_USER = new User(VALID_LOGIN1, INVALID_PWD, VALID_AGE);
+    private static final User INVALID_AGE_USER = new User(VALID_LOGIN1, VALID_PWD, INVALID_AGE);
+    private static final User VALID_USER = new User(VALID_LOGIN2, VALID_PWD, VALID_AGE);
+    private static final User VALID_USER2 = new User("Alexandr", "password", 50);
+    private static final User VALID_USER3 = new User("Leopold", "qwerty", 33);
+    private static RegistrationServiceImpl regService;
```
Comment (andrii-hoienko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:33]: Why static?

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+    @BeforeAll
+    static void beforeAll() {
```
Comment (andrii-hoienko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:36]: Why static?

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+                () -> {
+                regService.register(INVALID_AGE_USER);
+                });
+        assertEquals("Not valid age: "
+                + INVALID_AGE_USER.getAge()
+                + ". Min allowed age is " + MIN_AGE, invalidDataException.getMessage());
+    }
+
+    @Test
+    void register_oneUser_Ok() {
```
Comment (andrii-hoienko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:121]: Please make sure that user is stored in storage

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+    @Test
+    void register_twoUsers_Ok() {
+        User validUser2 = regService.register(VALID_USER2);
+        User validUser3 = regService.register(VALID_USER3);
+        assertEquals(validUser2, VALID_USER2);
+        assertEquals(validUser3, VALID_USER3);
+    }
+
+    @AfterAll
+    static void afterAll() {
```
Comment (andrii-hoienko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:135]: Better to clear storage before each to make all unit tests independant

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+                });
+        assertEquals("Not valid password length: "
+                + INVALID_PWD_USER.getPassword().length()
+                + ". Min allowed password length is "
+                + MIN_LENGTH, invalidDataException.getMessage());
+    }
+
+    @Test
+    void register_invalidAge_Exception_notOk() {
+        RegistrationException invalidDataException = assertThrows(RegistrationException.class,
```
Comment (andrii-hoienko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:111]: please refactor () -> {...} a bit, now it is hard to read

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+        assertEquals("Age can't be null", invalidDataException.getMessage());
+    }
+
+    @Test
+    void register_invalidLogin_Exception_notOk() {
+        RegistrationException invalidDataException = assertThrows(RegistrationException.class,
+                () -> {
+                regService.register(INVALID_LOGIN_USER);
+                });
+        assertEquals("Not valid login length: "
```
Comment (andrii-hoienko) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:82]: Please add some more cases for login and password lenght check, for example (0, 5, 3). 3 - is regular invalid case, 5 - edge case, 0 - edge case

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Nice, but some improvements required

**Review by Vasul-Pankiv — APPROVED** [review]
